### PR TITLE
[move-cli] add the option to show coverage with in move-cli test

### PR DESCRIPTION
**Pull request by mengxu-fb**
## Motivation

The Move CLI provides a `test` command for running expected value tests and a (separate) code coverage tool for measuring instruction coverage at the bytecode level. This task is to integrate these two tools so that `test` optionally produces a coverage report for the code under test. This will provide an awesome experience for new developers starting with Move: free code coverage info integrated with testing!

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Add an additional `--track-cov` flag to the end on how `move-cli test` is supposed to run, e.g.,
`cargo run --bin move-cli -- test tests/testsuite/module_run_view/args.txt --track-cov`
Coverage information will be printed, per module, per function.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/libra/developers.libra.org, and link to your PR here.)

### CODE REVIEW

**Review by tzakian — APPROVED** [review]
This looks amazing! This is going to make for an amazing CLI experience, and I can't wait to see this get in :) 

A couple future features that would be cool to implement using this info as well:
* Support for displaying coverage information for disassembled output of a module (it's already hooked up to the disassembler so would be pretty straightforward).
* Support for showing coverage on a source module, again I don't expect this to be that hard, but might be slightly more difficult than the disassembler.

**Comment by mengxu-fb** [comment]
Many thanks for the suggestion @tzakian! I'll keep the new feature implementation in the pipeline.

**Comment by mengxu-fb** [comment]
And also, I need someone to help land this PR for me :)

**Comment by tzakian** [comment]
Ah! sorry didn't realize.

/land

**Comment by mengxu-fb** [comment]
Test failed because `language/tools/move-cli/tests/cli_tests.rs` was added after this PR. Will do a rebase and fix this issue.

**Comment by mengxu-fb** [comment]
Re-requested review since I added several new changes, in particular,

1. Added a dummy test case called `move-cli/tests/metatests/cov` to show how the output should look like with and w/o coverage tracking info printed.

2. Added a logic to prevent the option `--track-cov` from being cascaded down the chain of `move-cli test` through `args.txt` files.

**Review by sblackshear — APPROVED** [review]
@dariorussi: would you mind accepting this so Meng can get credit for the bootcamp task? It only touches the CLI code (nothing that runs on the validator).

File: language/tools/move-cli/src/test.rs [hunk 1]
```diff
@@ -87,13 +127,34 @@ pub fn run_one(args_path: &Path, cli_binary: &str) -> anyhow::Result<()> {
...
         output += std::str::from_utf8(&cmd_output.stdout)?;
         output += std::str::from_utf8(&cmd_output.stderr)?;
+
+        // show coverage information
+        if track_cov && trace_file.exists() {
+            if !trace_file.exists() {
+                eprintln!("Trace file {} not found", trace_file.to_string_lossy());
+                eprintln!("Coverage is only applicable to the RUN command in args.txt");
+            } else {
+                show_coverage(Path::new(&trace_file), Path::new(&move_data))?;
```
[the remark this answers is not in the document]
Comment (mengxu-fb) [reply language/tools/move-cli/src/test.rs:162]: Thanks for the suggestion! That makes more sense. Let me quickly update this.

[the remark this answers is not in the document]
Comment (mengxu-fb) [reply language/tools/move-cli/src/test.rs:162]: Emm, it seems that not only `CoverageMap` but also a `Vec` of `CompiledModule` need to be returned from the function in order to print out the summary. This does not seem to fit in a small change... Shall I suggest another PR to handle this issue?

[the remark this answers is not in the document]
Comment (sblackshear) [reply language/tools/move-cli/src/test.rs:162]: My thinking was something like: return `Ok(CoverageMap::from_trace_file(trace_file))` here and call `show_coverage` (along with building the `CompiledModule` vec) at the end of `run_all` instead.

There's something interesting we'll eventually have to deal with here: test code is going to execute both the modules/scripts under test (whose coverage we care about and want to show) and test modules/scripts (whose coverage we don't care about and shouldn't show). We should probably think about how to organize tests to make it clear which modules/scripts fall into each category and do the right thing.

[the remark this answers is not in the document]
Comment (mengxu-fb) [reply language/tools/move-cli/src/test.rs:162]: I see your point. Yes, we should design a way to organize tests such that we have a control on which module we should show the coverage and which we should not.

[the remark this answers is not in the document]
Comment (mengxu-fb) [reply language/tools/move-cli/src/test.rs:162]: One way I can think of now is to have a `cov.cfg` (or `cov.txt`) file which serves as a blacklist/whitelist on modules we want to examine for coverage.

The overall process will be like:

1. Execute the commands in `args.txt`.
2. Collect whatever modules published under `move_data`.
3. Filter the module list through the `cov.cfg`.
4. For each module remaining, print its coverage information.

**Review by sblackshear — CHANGES_REQUESTED** [review]
This looks good to go, but we need to switch it from `libra:master` to `libra::feature_dev`

**Comment by mengxu-fb** [comment]
Thank you Sam! Let me do a rebase and create another PR onto `feature_dev`.

**Comment by sblackshear** [comment]
Superseded by https://github.com/libra/libra/pull/6486

**Pull request closed without merging**